src/include/hardware.h

```diff
@@ -31,19 +31,40 @@
  * same memory location, but to do this safely we need to precisely control how reads to memory are
  * performed. This macro gives us control over this.
  *
- * GCC and Clang have a __typeof__ compiler builtin which allows us to temporarily cast the value
- * being read as a volatile and achieve volatile semantics. For other compilers we'll fall back on
- * inserting a read barrier after the read (our pre-existing implementation) which prevents invented
- * and fused loads for this variable in the code following the expression.
+ * To allow this macro to be used in a backwards compatible way it has been implemented using
+ * barriers. Technically this can be achieved by a volatile cast.
+ */
+#define WT_READ_ONCE(v, val) \
+    do {                     \
+        WT_READ_BARRIER();   \
+        (v) = (val);         \
```

Comment:
I'll combine the reply to this comment as well here:
> "This does surprise me - adding a write barrier to the final write stopping prior "invented stores" isn't intuitive - though a lot of this isn't intuitive. The bi-directional barriers feel like overkill - could you help me understand when they are necessary?" 

> The volatile cast is gone - I presume that's deliberate, since a barrier stops the compiler applying any optimizations. Is that right?  

Yes that is correct, our barrier definition utilizes the memory clobber which effectively forms a compiler barrier. The GCC documentation has this to say about it: "To ensure memory contains correct values, GCC may need to flush specific register values to memory before executing the asm. Further, the compiler does not assume that any values read from memory before an asm remain unchanged after that asm; it reloads them as needed. Using the "memory" clobber effectively forms a read/write memory barrier for the compiler."

> The bi-directional barriers feel like overkill - could you help me understand when they are necessary?

For what its worth I'm happy to drop the bi-directional barriers. Effectively making `WT_READ_ONCE` a `WT_ORDERED_READ` and `WT_WRITE_ONCE` a `WT_PUBLISH`. Given in the read case the barrier comes after the read we can imagine a scenario where that isn't sufficient. 

```
1:x = memory
// some code here
2:x = memory
read_barrier
if (3:x > 0 && 4:x != 10) {
    // more code
}
```
Given this contrived example the first read from memory could be cached, and used for read 2, and then 3 and 4. Though I wouldn't expect this to happen in our code. I think perhaps the best option is to drop the bi-directional barriers and just go with what we have.